docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/DockerComposeRule.java

```diff
@@ -133,34 +152,78 @@ protected LogCollector logCollector() {
         return new DoNothingLogCollector();
     }
 
-    @Override
     public void before() throws IOException, InterruptedException {
         log.debug("Starting docker-compose cluster");
+
+        pullBuildAndUp();
+
+        logCollector().startCollecting(dockerCompose());
+
+        emitEventsFor().waitingForServices(this::waitForServices);
+    }
+
+    private void pullBuildAndUp() throws IOException, InterruptedException {
         if (pullOnStartup()) {
-            dockerCompose().pull();
+            emitEventsFor().pull(dockerCompose()::pull);
         }
 
-        dockerCompose().build();
+        emitEventsFor().build(dockerCompose()::build);
 
         DockerCompose upDockerCompose = dockerCompose();
         if (removeConflictingContainersOnStartup()) {
             upDockerCompose = new ConflictingContainerRemovingDockerCompose(upDockerCompose, docker());
         }
-        upDockerCompose.up();
 
-        logCollector().startCollecting(dockerCompose());
+        emitEventsFor().up(upDockerCompose::up);
+    }
+
+    private void waitForServices() throws InterruptedException {
         log.debug("Waiting for services");
-        new ClusterWait(ClusterHealthCheck.nativeHealthChecks(), nativeServiceHealthCheckTimeout())
-                .waitUntilReady(containers());
-        clusterWaits().forEach(clusterWait -> clusterWait.waitUntilReady(containers()));
+        Consumer<Cluster> nativeHealthCheckClusterWait =
+                emitEventsFor().nativeClusterWait(
+                        new ClusterWait(ClusterHealthCheck.nativeHealthChecks(), nativeServiceHealthCheckTimeout()));
+
+        List<Consumer<Cluster>> allClusterWaits = Stream.concat(
+                Stream.of(nativeHealthCheckClusterWait),
+                clusterWaits().stream().map(emitEventsFor()::userClusterWait))
+                .collect(Collectors.toList());
+
+        waitForAllClusterWaits(allClusterWaits);
+
         log.debug("docker-compose cluster started");
     }
 
-    @Override
+    private void waitForAllClusterWaits(List<Consumer<Cluster>> allClusterWaits) throws InterruptedException {
+        ListeningExecutorService executorService = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(
+                allClusterWaits.size(),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("dcr-wait-%d")
+                        .build()));
+
+        try {
+            ListenableFuture<?> listListenableFuture =
```

Comment:
why does this need to be done in parallel?

Reply:
Ditto for measuring how long it takes for each service to become healthy.